From f3216ae3353ecd3b5a9ea62faca397ec989142bd Mon Sep 17 00:00:00 2001
From: Li Wei <liwei825@huawei.com>
Date: Wed, 15 Feb 2023 16:51:22 +0800
Subject: [PATCH 324/921] [Huawei] ocfs2: journal adapt the adl module

Offering: HULK
ocfs2 inclusion
category: feature
bugzilla: NA

----------------------------------------

With the additional of adl, add customized modifications of journal.

Signed-off-by: Li Wei <liwei825@huawei.com>
---
 fs/ocfs2/journal.c | 484 ++++++++++++++++++++++++++++++++-------------
 fs/ocfs2/journal.h |  66 ++++++-
 2 files changed, 404 insertions(+), 146 deletions(-)

diff --git a/fs/ocfs2/journal.c b/fs/ocfs2/journal.c
index 0534800a472a..b6e29a48adba 100644
--- a/fs/ocfs2/journal.c
+++ b/fs/ocfs2/journal.c
@@ -25,7 +25,7 @@
 #include "alloc.h"
 #include "blockcheck.h"
 #include "dir.h"
-#include "dlmglue.h"
+#include "lockglue.h"
 #include "extent_map.h"
 #include "heartbeat.h"
 #include "inode.h"
@@ -42,9 +42,11 @@
 #include "buffer_head_io.h"
 #include "ocfs2_trace.h"
 
+#include "cluster/heartbeat.h"
+
 DEFINE_SPINLOCK(trans_inc_lock);
 
-#define ORPHAN_SCAN_SCHEDULE_TIMEOUT 300000
+#define ORPHAN_SCAN_SCHEDULE_TIMEOUT 1800000
 
 static int ocfs2_force_read_journal(struct inode *inode);
 static int ocfs2_recover_node(struct ocfs2_super *osb,
@@ -178,72 +180,73 @@ int ocfs2_recovery_init(struct ocfs2_super *osb)
 	osb->disable_recovery = 0;
 	osb->recovery_thread_task = NULL;
 	init_waitqueue_head(&osb->recovery_event);
+	init_waitqueue_head(&osb->ocfs2rec_event);
 
-	rm = kzalloc(sizeof(struct ocfs2_recovery_map) +
-		     osb->max_slots * sizeof(unsigned int),
-		     GFP_KERNEL);
+	rm = kzalloc(sizeof(struct ocfs2_recovery_map), GFP_KERNEL);
 	if (!rm) {
 		mlog_errno(-ENOMEM);
 		return -ENOMEM;
 	}
 
-	rm->rm_entries = (unsigned int *)((char *)rm +
-					  sizeof(struct ocfs2_recovery_map));
 	osb->recovery_map = rm;
 
 	return 0;
 }
 
-/* we can't grab the goofy sem lock from inside wait_event, so we use
- * memory barriers to make sure that we'll see the null task before
- * being woken up */
-static int ocfs2_recovery_thread_running(struct ocfs2_super *osb)
+static int ocfs2_launch_recovery_thread(struct ocfs2_super *osb)
+{
+	int status = 0;
+
+	mlog(0, "Starting ocfs2rec thread...\n");
+
+	osb->recovery_thread_task = kthread_run(__ocfs2_recovery_thread, osb,
+			"ocfs2rec-%s", osb->uuid_str);
+	if (IS_ERR(osb->recovery_thread_task)) {
+		status = PTR_ERR(osb->recovery_thread_task);
+		osb->recovery_thread_task = NULL;
+		mlog(ML_ERROR,
+			"%s: unable to launch ocfs2rec thread, error=%d\n",
+			osb->uuid_str, status);
+	}
+
+	return status;
+}
+
+static void ocfs2_complete_recovery_thread(struct ocfs2_super *osb)
 {
-	mb();
-	return osb->recovery_thread_task != NULL;
+	if (osb->recovery_thread_task) {
+		mlog(ML_KTHREAD, "Waiting for ocfs2rec thread to exit\n");
+		kthread_stop(osb->recovery_thread_task);
+		osb->recovery_thread_task = NULL;
+	}
 }
 
 void ocfs2_recovery_exit(struct ocfs2_super *osb)
 {
-	struct ocfs2_recovery_map *rm;
-
 	/* disable any new recovery threads and wait for any currently
 	 * running ones to exit. Do this before setting the vol_state. */
 	mutex_lock(&osb->recovery_lock);
 	osb->disable_recovery = 1;
 	mutex_unlock(&osb->recovery_lock);
-	wait_event(osb->recovery_event, !ocfs2_recovery_thread_running(osb));
+	ocfs2_complete_recovery_thread(osb);
+	/* kthread_stop can wait for ocfs2rec complete, so don't need to wait here. */
+	/* wait_event(osb->recovery_event, !ocfs2_recovery_thread_running(osb)); */
 
 	/* At this point, we know that no more recovery threads can be
 	 * launched, so wait for any recovery completion work to
 	 * complete. */
 	if (osb->ocfs2_wq)
 		flush_workqueue(osb->ocfs2_wq);
-
-	/*
-	 * Now that recovery is shut down, and the osb is about to be
-	 * freed,  the osb_lock is not taken here.
-	 */
-	rm = osb->recovery_map;
-	/* XXX: Should we bug if there are dirty entries? */
-
-	kfree(rm);
 }
 
 static int __ocfs2_recovery_map_test(struct ocfs2_super *osb,
 				     unsigned int node_num)
 {
-	int i;
 	struct ocfs2_recovery_map *rm = osb->recovery_map;
 
 	assert_spin_locked(&osb->osb_lock);
 
-	for (i = 0; i < rm->rm_used; i++) {
-		if (rm->rm_entries[i] == node_num)
-			return 1;
-	}
-
-	return 0;
+	return test_bit(node_num, rm->rm_nodes);
 }
 
 /* Behaves like test-and-set.  Returns the previous value */
@@ -259,10 +262,13 @@ static int ocfs2_recovery_map_set(struct ocfs2_super *osb,
 	}
 
 	/* XXX: Can this be exploited? Not from o2dlm... */
-	BUG_ON(rm->rm_used >= osb->max_slots);
+	BUG_ON(rm->rm_used >= O2NM_MAX_NODES);
 
-	rm->rm_entries[rm->rm_used] = node_num;
+	set_bit(node_num, rm->rm_nodes);
 	rm->rm_used++;
+
+	ocfs2_set_lowlevel_recovery_map(osb, node_num);
+
 	spin_unlock(&osb->osb_lock);
 
 	return 0;
@@ -271,23 +277,29 @@ static int ocfs2_recovery_map_set(struct ocfs2_super *osb,
 static void ocfs2_recovery_map_clear(struct ocfs2_super *osb,
 				     unsigned int node_num)
 {
-	int i;
 	struct ocfs2_recovery_map *rm = osb->recovery_map;
 
 	spin_lock(&osb->osb_lock);
-
-	for (i = 0; i < rm->rm_used; i++) {
-		if (rm->rm_entries[i] == node_num)
-			break;
+	if (!__ocfs2_recovery_map_test(osb, node_num)) {
+		spin_unlock(&osb->osb_lock);
+		return;
 	}
 
-	if (i < rm->rm_used) {
-		/* XXX: be careful with the pointer math */
-		memmove(&(rm->rm_entries[i]), &(rm->rm_entries[i + 1]),
-			(rm->rm_used - i - 1) * sizeof(unsigned int));
-		rm->rm_used--;
-	}
+	clear_bit(node_num, rm->rm_nodes);
+	rm->rm_used--;
+
+	ocfs2_clear_lowlevel_recovery_map(osb, node_num);
+
+	spin_unlock(&osb->osb_lock);
+}
 
+static void ocfs2_recovery_map_clear_all(struct ocfs2_super *osb)
+{
+	struct ocfs2_recovery_map *rm = osb->recovery_map;
+
+	spin_lock(&osb->osb_lock);
+	memset(rm->rm_nodes, 0, sizeof(rm->rm_nodes));
+	rm->rm_used = 0;
 	spin_unlock(&osb->osb_lock);
 }
 
@@ -314,7 +326,7 @@ static int ocfs2_commit_cache(struct ocfs2_super *osb)
 	jbd2_journal_unlock_updates(journal->j_journal);
 	if (status < 0) {
 		up_write(&journal->j_trans_barrier);
-		mlog_errno(status);
+		wake_up(&journal->j_checkpointed);
 		goto finally;
 	}
 
@@ -325,8 +337,9 @@ static int ocfs2_commit_cache(struct ocfs2_super *osb)
 	up_write(&journal->j_trans_barrier);
 
 	trace_ocfs2_commit_cache_end(journal->j_trans_id, flushed);
-
+#ifdef ENABLE_DLM
 	ocfs2_wake_downconvert_thread(osb);
+#endif
 	wake_up(&journal->j_checkpointed);
 finally:
 	return status;
@@ -337,11 +350,18 @@ handle_t *ocfs2_start_trans(struct ocfs2_super *osb, int max_buffs)
 	journal_t *journal = osb->journal->j_journal;
 	handle_t *handle;
 
+	BUG_ON(!osb);
+	if (ocfs2_recover_ro(osb))
+		return ERR_PTR(-EROFS);
+
 	BUG_ON(!osb || !osb->journal->j_journal);
 
 	if (ocfs2_is_hard_readonly(osb))
 		return ERR_PTR(-EROFS);
 
+	if (ocfs2_test_invalid_fs(osb))
+		return ERR_PTR(-EIO);
+
 	BUG_ON(osb->journal->j_state == OCFS2_JOURNAL_FREE);
 	BUG_ON(max_buffs <= 0);
 
@@ -483,7 +503,6 @@ int ocfs2_allocate_extend_trans(handle_t *handle, int thresh)
 	return status;
 }
 
-
 struct ocfs2_triggers {
 	struct jbd2_buffer_trigger_type	ot_triggers;
 	int				ot_offset;
@@ -552,14 +571,17 @@ static void ocfs2_db_frozen_trigger(struct jbd2_buffer_trigger_type *triggers,
 static void ocfs2_abort_trigger(struct jbd2_buffer_trigger_type *triggers,
 				struct buffer_head *bh)
 {
+	struct super_block *sb = bh->b_bdev->bd_super;
+
 	mlog(ML_ERROR,
 	     "ocfs2_abort_trigger called by JBD2.  bh = 0x%lx, "
 	     "bh->b_blocknr = %llu\n",
 	     (unsigned long)bh,
 	     (unsigned long long)bh->b_blocknr);
 
-	ocfs2_error(bh->b_bdev->bd_super,
-		    "JBD2 has aborted our journal, ocfs2 cannot continue\n");
+	/* journal is now aborted, set filesystem invalid */
+	if (!ocfs2_test_invalid_fs(OCFS2_SB(sb)))
+		o2hb_handle_invalid(OCFS2_SB(sb)->uuid_str);
 }
 
 static struct ocfs2_triggers di_triggers = {
@@ -646,31 +668,51 @@ static int __ocfs2_journal_access(handle_t *handle,
 	BUG_ON(!handle);
 	BUG_ON(!bh);
 
+	if (ocfs2_test_osb_flag(osb, OCFS2_OSB_INVALID_BUFFER)) {
+		mlog(ML_ERROR, "%s: OCFS2 found buffer already cached.\n",
+			osb->uuid_str);
+		return -EROFS;
+	}
+
+	if (ocfs2_test_invalid_fs(osb)) {
+		mlog(ML_ERROR, "%s: OCFS2 invalid because of disk timeout.\n",
+				osb->uuid_str);
+		return -EIO;
+	}
+
 	trace_ocfs2_journal_access(
 		(unsigned long long)ocfs2_metadata_cache_owner(ci),
 		(unsigned long long)bh->b_blocknr, type, bh->b_size);
 
 	/* we can safely remove this assertion after testing. */
 	if (!buffer_uptodate(bh)) {
-		mlog(ML_ERROR, "giving me a buffer that's not uptodate!\n");
-		mlog(ML_ERROR, "b_blocknr=%llu, b_state=0x%lx\n",
-		     (unsigned long long)bh->b_blocknr, bh->b_state);
+		mlog(ML_ERROR,
+			"%s: giving me a buffer that's not uptodate! b_blocknr=%llu, b_state = 0x%lx\n",
+			osb->uuid_str,
+			(unsigned long long)bh->b_blocknr, bh->b_state);
 
 		lock_buffer(bh);
 		/*
-		 * A previous transaction with a couple of buffer heads fail
-		 * to checkpoint, so all the bhs are marked as BH_Write_EIO.
-		 * For current transaction, the bh is just among those error
-		 * bhs which previous transaction handle. We can't just clear
-		 * its BH_Write_EIO and reuse directly, since other bhs are
-		 * not written to disk yet and that will cause metadata
-		 * inconsistency. So we should set fs read-only to avoid
-		 * further damage.
+		 * we should not reuse the dirty bh directly due to the
+		 * following situation:
+		 * when removing extent rec, we will dirty the bhs of extent rec
+		 * and truncate log at the same time, and hand them over to
+		 * journal. the bhs are not flushed to disk successfully due to
+		 * abnormal storage link. then the storage link become normal,
+		 * so truncate log flush worker found the dirty bh of truncate
+		 * log, and help flushing to disk successfully. the removing
+		 * space is still in extent rec, but truncate log flush worker
+		 * have given it back to globalalloc. that will cause duplicate
+		 * cluster problem. so we should set invalid in case of ruining
+		 * atomicity and consistency.
 		 */
 		if (buffer_write_io_error(bh) && !buffer_uptodate(bh)) {
+			mlog(ML_ERROR,
+				"A previous attempt to write this buffer head failed, b_blocknr=%llu, b_state=0x%lx",
+				(unsigned long long)bh->b_blocknr, bh->b_state);
+			o2hb_handle_invalid(osb->uuid_str);
 			unlock_buffer(bh);
-			return ocfs2_error(osb->sb, "A previous attempt to "
-					"write this buffer head failed\n");
+			return -EIO;
 		}
 		unlock_buffer(bh);
 	}
@@ -696,15 +738,19 @@ static int __ocfs2_journal_access(handle_t *handle,
 
 	default:
 		status = -EINVAL;
-		mlog(ML_ERROR, "Unknown access type!\n");
+		mlog(ML_ERROR, "%s: Unknown access type!\n", osb->uuid_str);
 	}
 	if (!status && ocfs2_meta_ecc(osb) && triggers)
 		jbd2_journal_set_triggers(bh, &triggers->ot_triggers);
 	ocfs2_metadata_cache_io_unlock(ci);
 
-	if (status < 0)
-		mlog(ML_ERROR, "Error %d getting %d access to buffer!\n",
-		     status, type);
+	if (status < 0) {
+		/* set EROFS only when journal aborted, convert to EIO */
+		if (status == -EROFS)
+			status = -EIO;
+		mlog(ML_ERROR, "%s: Error %d getting %d access to buffer!\n",
+		     osb->uuid_str, status, type);
+	}
 
 	return status;
 }
@@ -787,7 +833,7 @@ void ocfs2_journal_dirty(handle_t *handle, struct buffer_head *bh)
 					"Aborting transaction and journal.\n");
 			handle->h_err = status;
 			jbd2_journal_abort_handle(handle);
-			jbd2_journal_abort(journal, status);
+			jbd2_journal_abort(journal, -EIO);
 			ocfs2_abort(sb, "Journal already aborted.\n");
 		}
 	}
@@ -809,6 +855,9 @@ void ocfs2_set_journal_params(struct ocfs2_super *osb)
 		journal->j_flags |= JBD2_BARRIER;
 	else
 		journal->j_flags &= ~JBD2_BARRIER;
+
+	/* Abort the journal on file data write error in ordered mode */
+	journal->j_flags |= JBD2_ABORT_ON_SYNCDATA_ERR;
 	write_unlock(&journal->j_state_lock);
 }
 
@@ -835,7 +884,7 @@ int ocfs2_journal_init(struct ocfs2_journal *journal, int *dirty)
 		goto done;
 	}
 	if (is_bad_inode(inode)) {
-		mlog(ML_ERROR, "access error (bad inode)\n");
+		mlog(ML_ERROR, "%s: access error (bad inode)\n", osb->uuid_str);
 		iput(inode);
 		inode = NULL;
 		status = -EACCES;
@@ -851,16 +900,16 @@ int ocfs2_journal_init(struct ocfs2_journal *journal, int *dirty)
 	status = ocfs2_inode_lock_full(inode, &bh, 1, OCFS2_META_LOCK_RECOVERY);
 	if (status < 0) {
 		if (status != -ERESTARTSYS)
-			mlog(ML_ERROR, "Could not get lock on journal!\n");
+			mlog(ML_ERROR, "%s: Could not get lock on journal!\n", osb->uuid_str);
 		goto done;
 	}
 
 	inode_lock = 1;
 	di = (struct ocfs2_dinode *)bh->b_data;
 
-	if (i_size_read(inode) <  OCFS2_MIN_JOURNAL_SIZE) {
-		mlog(ML_ERROR, "Journal file size (%lld) is too small!\n",
-		     i_size_read(inode));
+	if (i_size_read(inode) < OCFS2_MIN_JOURNAL_SIZE) {
+		mlog(ML_ERROR, "%s: Journal file size (%lld) is too small!\n",
+		     osb->uuid_str, i_size_read(inode));
 		status = -EINVAL;
 		goto done;
 	}
@@ -872,7 +921,7 @@ int ocfs2_journal_init(struct ocfs2_journal *journal, int *dirty)
 	/* call the kernels journal init function now */
 	j_journal = jbd2_journal_init_inode(inode);
 	if (j_journal == NULL) {
-		mlog(ML_ERROR, "Linux journal layer error\n");
+		mlog(ML_ERROR, "%s: Linux journal layer error\n", osb->uuid_str);
 		status = -EINVAL;
 		goto done;
 	}
@@ -933,7 +982,15 @@ static int ocfs2_journal_toggle_dirty(struct ocfs2_super *osb,
 	/* The journal bh on the osb always comes from ocfs2_journal_init()
 	 * and was validated there inside ocfs2_inode_lock_full().  It's a
 	 * code bug if we mess it up. */
-	BUG_ON(!OCFS2_IS_VALID_DINODE(fe));
+	/* journal bh may be re-read in function ocfs2_check_journals_nolocks,
+	 * so the journal bh may be corrupted then */
+	if (!OCFS2_IS_VALID_DINODE(fe)) {
+		ocfs2_error(osb->sb, "Invalid dinode #%llu: signature = %.*s",
+				(unsigned long long)bh->b_blocknr, 7,
+				fe->i_signature);
+		status = -EROFS;
+		return status;
+	}
 
 	flags = le32_to_cpu(fe->id1.journal1.ij_flags);
 	if (dirty)
@@ -967,7 +1024,7 @@ void ocfs2_journal_shutdown(struct ocfs2_super *osb)
 	BUG_ON(!osb);
 
 	journal = osb->journal;
-	if (!journal)
+	if (!journal || ocfs2_recover_ro(osb))
 		goto done;
 
 	inode = journal->j_inode;
@@ -1014,7 +1071,14 @@ void ocfs2_journal_shutdown(struct ocfs2_super *osb)
 		 * Do not toggle if flush was unsuccessful otherwise
 		 * will leave dirty metadata in a "clean" journal
 		 */
-		status = ocfs2_journal_toggle_dirty(osb, 0, 0);
+		/* before journal destroy, note down osb state */
+		if (ocfs2_test_invalid_fs(osb) ||
+				is_journal_aborted(journal->j_journal)) {
+			mlog(ML_ERROR, "%s: OCFS2 need recovery.\n", osb->uuid_str);
+			status = ocfs2_journal_toggle_dirty(osb, 1, 0);
+		} else
+			status = ocfs2_journal_toggle_dirty(osb, 0, 0);
+
 		if (status < 0)
 			mlog_errno(status);
 	}
@@ -1043,10 +1107,11 @@ static void ocfs2_clear_journal_error(struct super_block *sb,
 
 	olderr = jbd2_journal_errno(journal);
 	if (olderr) {
-		mlog(ML_ERROR, "File system error %d recorded in "
-		     "journal %u.\n", olderr, slot);
-		mlog(ML_ERROR, "File system on device %s needs checking.\n",
-		     sb->s_id);
+		mlog(ML_ERROR,
+			"%s: File system error %d recorded in journal %u.\n",
+			OCFS2_SB(sb)->uuid_str, olderr, slot);
+		mlog(ML_ERROR, "%s: File system on device %s needs checking.\n",
+		    OCFS2_SB(sb)->uuid_str, sb->s_id);
 
 		jbd2_journal_ack_err(journal);
 		jbd2_journal_clear_err(journal);
@@ -1064,7 +1129,7 @@ int ocfs2_journal_load(struct ocfs2_journal *journal, int local, int replayed)
 
 	status = jbd2_journal_load(journal->j_journal);
 	if (status < 0) {
-		mlog(ML_ERROR, "Failed to load journal!\n");
+		mlog(ML_ERROR, "%s: Failed to load journal!\n", osb->uuid_str);
 		goto done;
 	}
 
@@ -1074,8 +1139,10 @@ int ocfs2_journal_load(struct ocfs2_journal *journal, int local, int replayed)
 		jbd2_journal_lock_updates(journal->j_journal);
 		status = jbd2_journal_flush(journal->j_journal);
 		jbd2_journal_unlock_updates(journal->j_journal);
-		if (status < 0)
+		if (status < 0) {
 			mlog_errno(status);
+			goto done;
+		}
 	}
 
 	status = ocfs2_journal_toggle_dirty(osb, 1, replayed);
@@ -1084,6 +1151,12 @@ int ocfs2_journal_load(struct ocfs2_journal *journal, int local, int replayed)
 		goto done;
 	}
 
+	status = ocfs2_launch_recovery_thread(osb);
+	if (status < 0) {
+		mlog_errno(status);
+		goto done;
+	}
+
 	/* Launch the commit thread */
 	if (!local) {
 		osb->commit_task = kthread_run(ocfs2_commit_thread, osb,
@@ -1091,8 +1164,10 @@ int ocfs2_journal_load(struct ocfs2_journal *journal, int local, int replayed)
 		if (IS_ERR(osb->commit_task)) {
 			status = PTR_ERR(osb->commit_task);
 			osb->commit_task = NULL;
-			mlog(ML_ERROR, "unable to launch ocfs2commit thread, "
-			     "error=%d", status);
+			ocfs2_complete_recovery_thread(osb);
+			mlog(ML_ERROR,
+				"%s: unable to launch ocfs2commit thread, error=%d\n",
+				osb->uuid_str, status);
 			goto done;
 		}
 	} else
@@ -1125,7 +1200,7 @@ int ocfs2_journal_wipe(struct ocfs2_journal *journal, int full)
 	return status;
 }
 
-static int ocfs2_recovery_completed(struct ocfs2_super *osb)
+int ocfs2_recovery_completed(struct ocfs2_super *osb)
 {
 	int empty;
 	struct ocfs2_recovery_map *rm = osb->recovery_map;
@@ -1137,9 +1212,24 @@ static int ocfs2_recovery_completed(struct ocfs2_super *osb)
 	return empty;
 }
 
-void ocfs2_wait_for_recovery(struct ocfs2_super *osb)
+int ocfs2_wait_for_recovery(struct ocfs2_super *osb)
 {
-	wait_event(osb->recovery_event, ocfs2_recovery_completed(osb));
+	int timeout;
+	int status = 0;
+
+	timeout = wait_event_timeout(osb->recovery_event,
+			ocfs2_recovery_completed(osb),
+			msecs_to_jiffies(ocfs2_lock_hung_check_ms(osb)));
+	if (!timeout) {
+		mlog(ML_ERROR, "%s: ocfs2rec thread last for %lu ms, setting invalid\n",
+				osb->uuid_str, ocfs2_lock_hung_check_ms(osb));
+
+		if (!ocfs2_test_invalid_fs(osb))
+			o2hb_handle_invalid(osb->uuid_str);
+		status = -EIO;
+	}
+
+	return status;
 }
 
 /*
@@ -1187,7 +1277,7 @@ static int ocfs2_force_read_journal(struct inode *inode)
 				mlog_errno(status);
 				goto bail;
 			}
-
+			clear_buffer_verified(bh);
 			brelse(bh);
 			bh = NULL;
 		}
@@ -1338,6 +1428,9 @@ void ocfs2_complete_mount_recovery(struct ocfs2_super *osb)
 	if (ocfs2_is_hard_readonly(osb))
 		return;
 
+	if (ocfs2_recover_ro(osb))
+		return;
+
 	/* No need to queue up our truncate_log as regular cleanup will catch
 	 * that */
 	ocfs2_queue_recovery_completion(journal, osb->slot_num,
@@ -1366,7 +1459,9 @@ void ocfs2_complete_quota_recovery(struct ocfs2_super *osb)
 	}
 }
 
-static int __ocfs2_recovery_thread(void *arg)
+#define OCFS2_RECOVERY_TIMEOUT_JIFFIES msecs_to_jiffies(2*60*1000)
+#define OCFS2_INVALID_NODE_NUM	2000
+static int __ocfs2_recovery_thread_do_work(void *arg)
 {
 	int status, node_num, slot_num;
 	struct ocfs2_super *osb = arg;
@@ -1374,6 +1469,7 @@ static int __ocfs2_recovery_thread(void *arg)
 	int *rm_quota = NULL;
 	int rm_quota_used = 0, i;
 	struct ocfs2_quota_recovery *qrec;
+	int exit = 0;
 
 	/* Whether the quota supported. */
 	int quota_enabled = OCFS2_HAS_RO_COMPAT_FEATURE(osb->sb,
@@ -1383,16 +1479,21 @@ static int __ocfs2_recovery_thread(void *arg)
 
 	status = ocfs2_wait_on_mount(osb);
 	if (status < 0) {
+		exit = 1;
 		goto bail;
 	}
 
+rezalloc:
 	if (quota_enabled) {
 		rm_quota = kcalloc(osb->max_slots, sizeof(int), GFP_NOFS);
 		if (!rm_quota) {
-			status = -ENOMEM;
-			goto bail;
+			mlog(ML_NOTICE, "%s: alloc memory failed, rezalloc after 50ms\n",
+					osb->uuid_str);
+			msleep(50);
+			goto rezalloc;
 		}
 	}
+
 restart:
 	status = ocfs2_super_lock(osb, 1);
 	if (status < 0) {
@@ -1412,11 +1513,12 @@ static int __ocfs2_recovery_thread(void *arg)
 	while (rm->rm_used) {
 		/* It's always safe to remove entry zero, as we won't
 		 * clear it until ocfs2_recover_node() has succeeded. */
-		node_num = rm->rm_entries[0];
+		node_num = find_first_bit(rm->rm_nodes, O2NM_MAX_NODES);
 		spin_unlock(&osb->osb_lock);
 		slot_num = ocfs2_node_num_to_slot(osb, node_num);
 		trace_ocfs2_recovery_thread_node(node_num, slot_num);
-		if (slot_num == -ENOENT) {
+
+		if (slot_num == -ENOENT || ocfs2_test_invalid_fs(osb)) {
 			status = 0;
 			goto skip_recovery;
 		}
@@ -1428,10 +1530,11 @@ static int __ocfs2_recovery_thread(void *arg)
 		 * the slot. So we remember which nodes need quota recovery
 		 * and when everything else is done, we recover quotas. */
 		if (quota_enabled) {
-			for (i = 0; i < rm_quota_used
-					&& rm_quota[i] != slot_num; i++)
-				;
-
+			for (i = 0; i < rm_quota_used; i++) {
+				if (rm_quota[i] == slot_num) {
+					break;
+				};
+			}
 			if (i == rm_quota_used)
 				rm_quota[rm_quota_used++] = slot_num;
 		}
@@ -1441,13 +1544,15 @@ static int __ocfs2_recovery_thread(void *arg)
 		if (!status) {
 			ocfs2_recovery_map_clear(osb, node_num);
 		} else {
-			mlog(ML_ERROR,
-			     "Error %d recovering node %d on device (%u,%u)!\n",
-			     status, node_num,
-			     MAJOR(osb->sb->s_dev), MINOR(osb->sb->s_dev));
-			mlog(ML_ERROR, "Volume requires unmount.\n");
+			static unsigned long warn_time;
+
+			if (printk_timed_ratelimit(&warn_time, 60*HZ))
+				mlog(ML_ERROR,
+					"%s: Error %d recovering node %d on device(%u,%u)!\n",
+					osb->uuid_str, status, node_num,
+					MAJOR(osb->sb->s_dev), MINOR(osb->sb->s_dev));
+			msleep(20);
 		}
-
 		spin_lock(&osb->osb_lock);
 	}
 	spin_unlock(&osb->osb_lock);
@@ -1483,15 +1588,18 @@ static int __ocfs2_recovery_thread(void *arg)
 	ocfs2_queue_replay_slots(osb, ORPHAN_NEED_TRUNCATE);
 
 bail:
+	if (ocfs2_test_invalid_fs(osb) || exit)
+		ocfs2_recovery_map_clear_all(osb);
+
 	mutex_lock(&osb->recovery_lock);
-	if (!status && !ocfs2_recovery_completed(osb)) {
+	if (!ocfs2_recovery_completed(osb)) {
 		mutex_unlock(&osb->recovery_lock);
+		if (status)
+			msleep(2000);
 		goto restart;
 	}
 
 	ocfs2_free_replay_slots(osb);
-	osb->recovery_thread_task = NULL;
-	mb(); /* sync with ocfs2_recovery_thread_running */
 	wake_up(&osb->recovery_event);
 
 	mutex_unlock(&osb->recovery_lock);
@@ -1502,7 +1610,27 @@ static int __ocfs2_recovery_thread(void *arg)
 	/* no one is callint kthread_stop() for us so the kthread() api
 	 * requires that we call do_exit().  And it isn't exported, but
 	 * complete_and_exit() seems to be a minimal wrapper around it. */
-	complete_and_exit(NULL, status);
+
+	/* complete_and_exit(NULL, status); */
+	return status;
+}
+
+static int __ocfs2_recovery_thread(void *arg)
+{
+	int status = 0;
+	struct ocfs2_super *osb = arg;
+
+	while (!(kthread_should_stop() && ocfs2_recovery_completed(osb))) {
+		wait_event_interruptible(osb->ocfs2rec_event,
+				!ocfs2_recovery_completed(osb) || kthread_should_stop());
+
+		if (!ocfs2_recovery_completed(osb)) {
+			mlog(ML_NOTICE, "%s: begin to do ocfs2 recovery\n", osb->uuid_str);
+			__ocfs2_recovery_thread_do_work(osb);
+		}
+	}
+
+	return status;
 }
 
 void ocfs2_recovery_thread(struct ocfs2_super *osb, int node_num)
@@ -1516,16 +1644,9 @@ void ocfs2_recovery_thread(struct ocfs2_super *osb, int node_num)
 
 	if (osb->disable_recovery)
 		goto out;
-
+	/* wakeup ocfs2rec thread */
 	if (osb->recovery_thread_task)
-		goto out;
-
-	osb->recovery_thread_task =  kthread_run(__ocfs2_recovery_thread, osb,
-			"ocfs2rec-%s", osb->uuid_str);
-	if (IS_ERR(osb->recovery_thread_task)) {
-		mlog_errno((int)PTR_ERR(osb->recovery_thread_task));
-		osb->recovery_thread_task = NULL;
-	}
+		wake_up(&osb->ocfs2rec_event);
 
 out:
 	mutex_unlock(&osb->recovery_lock);
@@ -1542,9 +1663,11 @@ static int ocfs2_read_journal_inode(struct ocfs2_super *osb,
 
 	BUG_ON(slot_num >= osb->max_slots);
 
-	inode = ocfs2_get_system_file_inode(osb, JOURNAL_SYSTEM_INODE,
+	inode = _ocfs2_get_system_file_inode(osb, JOURNAL_SYSTEM_INODE,
 					    slot_num);
 	if (!inode || is_bad_inode(inode)) {
+		if (ocfs2_is_soft_readonly(osb))
+			status = -EINVAL;
 		mlog_errno(status);
 		goto bail;
 	}
@@ -1586,6 +1709,9 @@ static int ocfs2_replay_journal(struct ocfs2_super *osb,
 	status = ocfs2_read_journal_inode(osb, slot_num, &bh, &inode);
 	if (status) {
 		mlog_errno(status);
+		/* set file system invalid when journal invalid */
+		if (status == -EINVAL)
+			o2hb_handle_invalid(osb->uuid_str);
 		goto done;
 	}
 
@@ -1610,13 +1736,32 @@ static int ocfs2_replay_journal(struct ocfs2_super *osb,
 		goto done;
 	}
 
+	/*
+	 * There is a small chance that another node has already recovered
+	 * the slot, and then the slot has been owned by node after
+	 * successfully mounted, if we recover this slot again, it will be
+	 * blocked when get journal inode lock, so just return busy status
+	 */
+	status = ocfs2_trylock_journal(osb, slot_num);
+	if (status < 0) {
+		if (status == -EAGAIN) {
+			status = -EBUSY;
+			mlog(ML_NOTICE,
+				"%s: This slot(%d) has already been owned, we should not recovery it\n",
+				osb->uuid_str, slot_num);
+			goto done;
+		}
+		mlog_errno(status);
+		goto done;
+	}
+
 	/* Continue with recovery as the journal has not yet been recovered */
 
 	status = ocfs2_inode_lock_full(inode, &bh, 1, OCFS2_META_LOCK_RECOVERY);
 	if (status < 0) {
 		trace_ocfs2_replay_journal_lock_err(status);
 		if (status != -ERESTARTSYS)
-			mlog(ML_ERROR, "Could not lock journal!\n");
+			mlog(ML_ERROR, "%s: Could not lock journal!\n", osb->uuid_str);
 		goto done;
 	}
 	got_lock = 1;
@@ -1636,21 +1781,23 @@ static int ocfs2_replay_journal(struct ocfs2_super *osb,
 	/* we need to run complete recovery for offline orphan slots */
 	ocfs2_replay_map_set_state(osb, REPLAY_NEEDED);
 
-	printk(KERN_NOTICE "ocfs2: Begin replay journal (node %d, slot %d) on "\
-	       "device (%u,%u)\n", node_num, slot_num, MAJOR(osb->sb->s_dev),
-	       MINOR(osb->sb->s_dev));
+	mlog(ML_NOTICE, "%s: Recovering node %d from slot %d on device (%u,%u)\n",
+	     osb->uuid_str, node_num, slot_num,
+	     MAJOR(osb->sb->s_dev), MINOR(osb->sb->s_dev));
 
 	OCFS2_I(inode)->ip_clusters = le32_to_cpu(fe->i_clusters);
 
 	status = ocfs2_force_read_journal(inode);
 	if (status < 0) {
 		mlog_errno(status);
+		if (status == -EINVAL)
+			o2hb_handle_invalid(osb->uuid_str);
 		goto done;
 	}
 
 	journal = jbd2_journal_init_inode(inode);
 	if (journal == NULL) {
-		mlog(ML_ERROR, "Linux journal layer error\n");
+		mlog(ML_ERROR, "%s: Linux journal layer error\n", osb->uuid_str);
 		status = -EIO;
 		goto done;
 	}
@@ -1660,6 +1807,9 @@ static int ocfs2_replay_journal(struct ocfs2_super *osb,
 		mlog_errno(status);
 		if (!igrab(inode))
 			BUG();
+		/* set file system soft_readonly when journal invalid */
+		if (status == -EINVAL)
+			ocfs2_handle_recovery_invalid(osb);
 		jbd2_journal_destroy(journal);
 		goto done;
 	}
@@ -1693,9 +1843,6 @@ static int ocfs2_replay_journal(struct ocfs2_super *osb,
 
 	jbd2_journal_destroy(journal);
 
-	printk(KERN_NOTICE "ocfs2: End replay journal (node %d, slot %d) on "\
-	       "device (%u,%u)\n", node_num, slot_num, MAJOR(osb->sb->s_dev),
-	       MINOR(osb->sb->s_dev));
 done:
 	/* drop the lock on this nodes journal */
 	if (got_lock)
@@ -1747,6 +1894,8 @@ static int ocfs2_recover_node(struct ocfs2_super *osb,
 	status = ocfs2_begin_local_alloc_recovery(osb, slot_num, &la_copy);
 	if (status < 0) {
 		mlog_errno(status);
+		if (status == -EINVAL)
+			o2hb_handle_invalid(osb->uuid_str);
 		goto done;
 	}
 
@@ -1782,15 +1931,15 @@ static int ocfs2_trylock_journal(struct ocfs2_super *osb,
 	int status, flags;
 	struct inode *inode = NULL;
 
-	inode = ocfs2_get_system_file_inode(osb, JOURNAL_SYSTEM_INODE,
+	inode = _ocfs2_get_system_file_inode(osb, JOURNAL_SYSTEM_INODE,
 					    slot_num);
 	if (inode == NULL) {
-		mlog(ML_ERROR, "access error\n");
+		mlog(ML_ERROR, "%s: access error\n", osb->uuid_str);
 		status = -EACCES;
 		goto bail;
 	}
 	if (is_bad_inode(inode)) {
-		mlog(ML_ERROR, "access error (bad inode)\n");
+		mlog(ML_ERROR, "%s: access error (bad inode)\n", osb->uuid_str);
 		iput(inode);
 		inode = NULL;
 		status = -EACCES;
@@ -1850,9 +1999,11 @@ int ocfs2_mark_dead_nodes(struct ocfs2_super *osb)
 		}
 
 		status = ocfs2_slot_to_node_num_locked(osb, i, &node_num);
-		if (status == -ENOENT) {
+		if (status < 0) {
 			spin_unlock(&osb->osb_lock);
-			continue;
+			if (status == -ENOENT)
+				continue;
+			goto bail;
 		}
 
 		if (__ocfs2_recovery_map_test(osb, node_num)) {
@@ -1866,6 +2017,22 @@ int ocfs2_mark_dead_nodes(struct ocfs2_super *osb)
 		 * file here to test if he's alive. */
 		status = ocfs2_trylock_journal(osb, i);
 		if (!status) {
+			/* the dead node perhaps has not been fenced in
+			 * two cases:
+			 * 1.the dead node is the last one in domain, no one
+			 *   will help fencing it;
+			 * 2.the dead node become invalid very fast due to
+			 *   journal aborted, other nodes have not caught it.
+			 *
+			 * so, we need fence the dead node before recoverying
+			 * it, in case that dead node will re-write data on disk. */
+			if (!o2hb_is_in_domain(osb->uuid_str, node_num)) {
+				mlog(ML_NOTICE,
+					"%s: fence dead node %d before recoverying it\n",
+					osb->uuid_str, node_num);
+				ocfs2_fence_scsi(NULL, node_num, osb, 0);
+			}
+
 			/* Since we're called from mount, we know that
 			 * the recovery thread can't race us on
 			 * setting / checking the recovery bits. */
@@ -1948,10 +2115,15 @@ static void ocfs2_queue_orphan_scan(struct ocfs2_super *osb)
 		goto unlock;
 
 	if (os->os_seqno != seqno) {
+		mlog(ML_NOTICE, "%s: orphan scan lock info, os_seqno %u, seqno %u\n",
+				osb->uuid_str, os->os_seqno, seqno);
 		os->os_seqno = seqno;
 		goto unlock;
 	}
 
+	if (ocfs2_test_invalid_fs(osb))
+		goto unlock;
+
 	for (i = 0; i < osb->max_slots; i++)
 		ocfs2_queue_recovery_completion(osb->journal, i, NULL, NULL,
 						NULL, ORPHAN_NO_NEED_TRUNCATE);
@@ -2019,7 +2191,8 @@ void ocfs2_orphan_scan_start(struct ocfs2_super *osb)
 
 	os = &osb->osb_orphan_scan;
 	os->os_scantime = ktime_get_seconds();
-	if (ocfs2_is_hard_readonly(osb) || ocfs2_mount_local(osb))
+	if (ocfs2_is_hard_readonly(osb) || ocfs2_mount_local(osb) ||
+			ocfs2_recover_ro(osb))
 		atomic_set(&os->os_state, ORPHAN_SCAN_INACTIVE);
 	else {
 		atomic_set(&os->os_state, ORPHAN_SCAN_ACTIVE);
@@ -2032,6 +2205,8 @@ struct ocfs2_orphan_filldir_priv {
 	struct dir_context	ctx;
 	struct inode		*head;
 	struct ocfs2_super	*osb;
+	struct inode		*orphan_dir_inode;
+	struct buffer_head	*orphan_dir_bh;
 	enum ocfs2_orphan_reco_type orphan_reco_type;
 };
 
@@ -2042,6 +2217,7 @@ static int ocfs2_orphan_filldir(struct dir_context *ctx, const char *name,
 	struct ocfs2_orphan_filldir_priv *p =
 		container_of(ctx, struct ocfs2_orphan_filldir_priv, ctx);
 	struct inode *iter;
+	int status;
 
 	if (name_len == 1 && !strncmp(".", name, 1))
 		return 0;
@@ -2057,8 +2233,23 @@ static int ocfs2_orphan_filldir(struct dir_context *ctx, const char *name,
 	/* Skip bad inodes so that recovery can continue */
 	iter = ocfs2_iget(p->osb, ino,
 			  OCFS2_FI_FLAG_ORPHAN_RECOVERY, 0);
-	if (IS_ERR(iter))
+	if (IS_ERR(iter)) {
+		if (PTR_ERR(iter) != -EROFS) {
+			mlog(ML_ERROR,
+				"orphan_filldir() iget failed. ret %ld, name: %.*s, inode: %llu\n",
+				PTR_ERR(iter),
+				name_len, name, ino);
+			return 0;
+		}
+		mlog(ML_ERROR,
+			"Invalid dinode %llu found in orphan, force delete it\n",
+			ino);
+		status = ocfs2_remove_invalid_entry(p->osb, ino,
+				p->orphan_dir_inode);
+		if (status)
+			mlog_errno(status);
 		return 0;
+	}
 
 	if (!strncmp(name, OCFS2_DIO_ORPHAN_PREFIX,
 			OCFS2_DIO_ORPHAN_PREFIX_LEN))
@@ -2087,6 +2278,7 @@ static int ocfs2_queue_orphans(struct ocfs2_super *osb,
 {
 	int status;
 	struct inode *orphan_dir_inode = NULL;
+	struct buffer_head *orphan_dir_bh = NULL;
 	struct ocfs2_orphan_filldir_priv priv = {
 		.ctx.actor = ocfs2_orphan_filldir,
 		.osb = osb,
@@ -2094,6 +2286,8 @@ static int ocfs2_queue_orphans(struct ocfs2_super *osb,
 		.orphan_reco_type = orphan_reco_type
 	};
 
+	priv.orphan_reco_type = orphan_reco_type;
+
 	orphan_dir_inode = ocfs2_get_system_file_inode(osb,
 						       ORPHAN_DIR_SYSTEM_INODE,
 						       slot);
@@ -2104,12 +2298,14 @@ static int ocfs2_queue_orphans(struct ocfs2_super *osb,
 	}
 
 	inode_lock(orphan_dir_inode);
-	status = ocfs2_inode_lock(orphan_dir_inode, NULL, 0);
+	status = ocfs2_inode_lock(orphan_dir_inode, &orphan_dir_bh, 1);
 	if (status < 0) {
 		mlog_errno(status);
 		goto out;
 	}
 
+	priv.orphan_dir_inode = orphan_dir_inode;
+	priv.orphan_dir_bh = orphan_dir_bh;
 	status = ocfs2_dir_foreach(orphan_dir_inode, &priv.ctx);
 	if (status) {
 		mlog_errno(status);
@@ -2119,9 +2315,10 @@ static int ocfs2_queue_orphans(struct ocfs2_super *osb,
 	*head = priv.head;
 
 out_cluster:
-	ocfs2_inode_unlock(orphan_dir_inode, 0);
+	ocfs2_inode_unlock(orphan_dir_inode, 1);
 out:
 	inode_unlock(orphan_dir_inode);
+	brelse(orphan_dir_bh);
 	iput(orphan_dir_inode);
 	return status;
 }
@@ -2311,16 +2508,21 @@ static int ocfs2_commit_thread(void *arg)
 		status = ocfs2_commit_cache(osb);
 		if (status < 0) {
 			static unsigned long abort_warn_time;
-
 			/* Warn about this once per minute */
 			if (printk_timed_ratelimit(&abort_warn_time, 60*HZ))
-				mlog(ML_ERROR, "status = %d, journal is "
-						"already aborted.\n", status);
+				mlog(ML_ERROR,
+					"%s: status = %d, journal is already aborted.\n",
+					osb->uuid_str, status);
+
+			/* set filesystem invalid */
+			if (!ocfs2_test_invalid_fs(osb))
+				o2hb_handle_invalid(osb->uuid_str);
 			/*
 			 * After ocfs2_commit_cache() fails, j_num_trans has a
 			 * non-zero value.  Sleep here to avoid a busy-wait
 			 * loop.
 			 */
+
 			msleep_interruptible(1000);
 		}
 
@@ -2330,6 +2532,12 @@ static int ocfs2_commit_thread(void *arg)
 			     "shutdown\n",
 			     atomic_read(&journal->j_num_trans));
 		}
+
+		if (status < 0 && kthread_should_stop()) {
+			/* journal is now aborted, ignore current transactions and exit */
+			atomic_set(&journal->j_num_trans, 0);
+			break;
+		}
 	}
 
 	return 0;
diff --git a/fs/ocfs2/journal.h b/fs/ocfs2/journal.h
index eb7a21bac71e..2e6723cea542 100644
--- a/fs/ocfs2/journal.h
+++ b/fs/ocfs2/journal.h
@@ -31,7 +31,7 @@ struct ocfs2_dinode;
 
 struct ocfs2_recovery_map {
 	unsigned int rm_used;
-	unsigned int *rm_entries;
+	unsigned long rm_nodes[BITS_TO_LONGS(O2NM_MAX_NODES)];
 };
 
 
@@ -146,7 +146,9 @@ void ocfs2_orphan_scan_start(struct ocfs2_super *osb);
 void ocfs2_orphan_scan_stop(struct ocfs2_super *osb);
 
 void ocfs2_complete_recovery(struct work_struct *work);
-void ocfs2_wait_for_recovery(struct ocfs2_super *osb);
+int ocfs2_wait_for_recovery(struct ocfs2_super *osb);
+
+int ocfs2_recovery_completed(struct ocfs2_super *osb);
 
 int ocfs2_recovery_init(struct ocfs2_super *osb);
 void ocfs2_recovery_exit(struct ocfs2_super *osb);
@@ -184,6 +186,31 @@ int    ocfs2_mark_dead_nodes(struct ocfs2_super *osb);
 void   ocfs2_complete_mount_recovery(struct ocfs2_super *osb);
 void ocfs2_complete_quota_recovery(struct ocfs2_super *osb);
 
+static inline void ocfs2_set_journal_error(struct ocfs2_super *osb, int err)
+{
+	if (!osb->journal || !osb->journal->j_journal)
+		return;
+
+	if (!osb->journal->j_journal->j_errno)
+		osb->journal->j_journal->j_errno = err;
+}
+
+static inline void ocfs2_set_journal_abort(struct ocfs2_super *osb)
+{
+	journal_t *journal;
+
+	if (!osb->journal || !osb->journal->j_journal)
+		return;
+
+	journal = osb->journal->j_journal;
+	printk(KERN_ERR
+		"OCFS2: Aborting journal on device %s.\n", journal->j_devname);
+
+	write_lock(&journal->j_state_lock);
+	journal->j_flags |= JBD2_ABORT;
+	write_unlock(&journal->j_state_lock);
+}
+
 static inline void ocfs2_start_checkpoint(struct ocfs2_super *osb)
 {
 	wake_up(&osb->checkpoint_event);
@@ -193,7 +220,11 @@ static inline void ocfs2_checkpoint_inode(struct inode *inode)
 {
 	struct ocfs2_super *osb = OCFS2_SB(inode->i_sb);
 
-	if (ocfs2_mount_local(osb))
+	if (ocfs2_mount_local(osb) || ocfs2_recover_ro(osb))
+		return;
+
+	/* in ocfs2_journal_shutdown, before journal destroy we set it */
+	if (ocfs2_test_invalid_fs(osb))
 		return;
 
 	if (!ocfs2_ci_fully_checkpointed(INODE_CACHE(inode))) {
@@ -205,8 +236,12 @@ static inline void ocfs2_checkpoint_inode(struct inode *inode)
 		 * OK. */
 		ocfs2_start_checkpoint(osb);
 
+		/* check ocfs2 invalid or journal abort, if so, should wake up */
 		wait_event(osb->journal->j_checkpointed,
-			   ocfs2_ci_fully_checkpointed(INODE_CACHE(inode)));
+			ocfs2_ci_fully_checkpointed(INODE_CACHE(inode)) ||
+			ocfs2_test_invalid_fs(osb) ||
+			(osb->journal->j_journal ?
+				is_journal_aborted(osb->journal->j_journal) : 1));
 	}
 }
 
@@ -413,11 +448,17 @@ static inline int ocfs2_add_dir_index_credits(struct super_block *sb)
 		ocfs2_clusters_to_blocks(sb, 1);
 }
 
+/* lock alloc dinode + group desc + prev group desc if we relink */
+#define OCFS2_LOCKALLOC_ALLOC 3
+
+/* lock alloc dinode + group desc update. We don't relink on free yet. */
+#define OCFS2_LOCKALLOC_FREE 2
+
 /* parent fe, parent block, new file entry, index leaf, inode alloc fe, inode
  * alloc group descriptor + mkdir/symlink blocks + dir blocks + xattr
  * blocks + quota update */
 static inline int ocfs2_mknod_credits(struct super_block *sb, int is_dir,
-				      int xattr_credits)
+				      int xattr_credits, bool disk_lock)
 {
 	int dir_credits = OCFS2_DIR_LINK_ADDITIONAL_CREDITS;
 
@@ -425,7 +466,8 @@ static inline int ocfs2_mknod_credits(struct super_block *sb, int is_dir,
 		dir_credits += ocfs2_add_dir_index_credits(sb);
 
 	return 4 + OCFS2_SUBALLOC_ALLOC + dir_credits + xattr_credits +
-	       ocfs2_quota_trans_credits(sb);
+	       ocfs2_quota_trans_credits(sb) +
+	       (disk_lock ? OCFS2_LOCKALLOC_ALLOC : 0);
 }
 
 /* local alloc metadata change + main bitmap updates */
@@ -458,6 +500,13 @@ static inline int ocfs2_unlink_credits(struct super_block *sb)
  * orphan dir index leaf */
 #define OCFS2_DELETE_INODE_CREDITS (3 * OCFS2_INODE_UPDATE_CREDITS + 4)
 
+static inline int ocfs2_remove_inode_credits(struct super_block *sb,
+		bool disk_lock)
+{
+	return OCFS2_DELETE_INODE_CREDITS + ocfs2_quota_trans_credits(sb) +
+			(disk_lock ? OCFS2_LOCKALLOC_FREE : 0);
+}
+
 /* dinode + orphan dir dinode + extent tree leaf block + orphan dir entry +
  * orphan dir index root + orphan dir index leaf */
 #define OCFS2_INODE_ADD_TO_ORPHAN_CREDITS  (2 * OCFS2_INODE_UPDATE_CREDITS + 4)
@@ -541,9 +590,10 @@ static inline int ocfs2_calc_extend_credits(struct super_block *sb,
 	       ocfs2_quota_trans_credits(sb);
 }
 
-static inline int ocfs2_calc_symlink_credits(struct super_block *sb)
+static inline int ocfs2_calc_symlink_credits(struct super_block *sb,
+		bool disk_lock)
 {
-	int blocks = ocfs2_mknod_credits(sb, 0, 0);
+	int blocks = ocfs2_mknod_credits(sb, 0, 0, disk_lock);
 
 	/* links can be longer than one block so we may update many
 	 * within our single allocated extent. */
-- 
2.17.1

